nucypher/device/trezor.py

```diff
@@ -0,0 +1,178 @@
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+nucypher is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+import rlp
+from functools import wraps
+from typing import Tuple
+
+from eth_account._utils.transactions import assert_valid_fields, Transaction
+from eth_utils import apply_key_map, to_canonical_address
+from nucypher.blockchain.eth.decorators import validate_checksum_address
+from nucypher.crypto.signing import InvalidSignature
+from nucypher.device.base import TrustedDevice
+
+try:
+    from trezorlib import client as trezor_client
+    from trezorlib import device as trezor_device
+    from trezorlib import ethereum as trezor_eth
+    from trezorlib import tools as trezor_tools
+    from trezorlib.transport import TransportException
+    from usb1 import USBErrorNoDevice, USBErrorBusy
+except ImportError:
+    raise RuntimeError("The nucypher package wasn't installed with the 'trezor' extra")
+
+
+class Trezor(TrustedDevice):
+    """
+    An implementation of a Trezor device for staking on the NuCypher network.
+    """
+
+    def __init__(self, cached_index: int = 1):
+        """
+        Initializes a Trezor device.
+
+        `cached_index` is the highest address_index to cache in memory.
+        TODO: Use a device config to determine some init settings like how many
+              addresses to cache.
+        TODO: We can potentially use the `_handle_device_call` decorator to
+              perform some input validation on each parameter in the calls.
+        """
+        try:
+            self.client = trezor_client.get_default_client()
+        except TransportException:
+            raise RuntimeError("Could not find a TREZOR device to connect to. Have you unlocked it?")
+
+        self.__addresses = {}
+        for addr_idx in range(cached_index + 1):
+            hd_path = self.ETH_BIP44_PATH.format(address_index=addr_idx)
+            address = self.get_address(hd_path=hd_path)
+            self.__addresses[address] = trezor_tools.parse_path(hd_path)
+
+    def _handle_device_call(device_func):
+        @wraps(device_func)
+        def wrapped_call(inst, *args, **kwargs):
+            try:
+                result = device_func(inst, *args, **kwargs)
+            except USBErrorNoDevice:
+                raise inst.DeviceError("The client cannot communicate to the TREZOR USB device. Was it disconnected?")
+            except USBErrorBusy:
+                raise inst.DeviceError("The TREZOR USB device is busy.")
+            else:
+                return result
+        return wrapped_call
```

Comment:
This is a cool decorator 😎 Just a question, why making it "private"?

Reply:
Shouldn't be used outside of `TrusteDevice` subclasses, imho.